### PR TITLE
Make shard allocation give preference to current location

### PR DESCRIPTION
**Pull request by dannyroberts**
In this implementation, it gives infinite preference to the current location.
This works well in most if not all recent migrations that I am aware of,
but will work poorly in the special case of simply adding a new node,
without removing any nodes or increasing the number of copies. In this case
all shards will stay where they are and no shards will end up on the new nodes.

I think this is still a huge practical improvement over the former algorithm.
In the future if this becomes a concern, I think a potential solution is basically
to do this first, and then go through a secondary rebalancing step,
where the preference for current location is loosened and shards are moved from
fuller nodes to emptier nodes until an acceptable uniformity is reached.

Update: I implemented rebalancing when there's an extreme disparity—mainly useful for when adding a new node. It still makes sure that the majority of the copies of any given shard don't move, which is important for data consistency if you're doing it live.

### CODE REVIEW

**Comment by emord** [comment]
> but will work poorly in the special case of simply adding a new node,

From what I know, this is the only expected migration expected to take place in the next few months (perhaps on ICDS if expanding the cluster). Are there any migrations that you expect to occur that would benefit from this?

**Comment by dannyroberts** [comment]
Oh, is that so? I'm planning on moving shards from couch[0-2]-production to
couch[3-10]-production (in multiple steps so there are pivot shards). For
that I want it to keep the same allocation for all existing commcarehq
(main DB) shards, and also keep locations of shards added over the multiple
steps without churn.

Sounds like you're saying that sometime soon we plan to add nodes on icds
have all existing shards get evenly rebalanced across the old and new
nodes. Is that right?

**Comment by dannyroberts** [comment]
@snopoke I think the changes I'm making here are still worthwhile. I'm going to add something that makes the behavior reasonable in the situation where you're adding an empty node and want shards redistributed to it.

About your specific issue with the user db, I think if you find something like that happening, you can now (with the changes I made in a recent PR) handle that by calling out the user db separately; then the allocation for the rest of the shards won't take it into account. If you have nodes couch0 through couch5, you could do:

```
target_allocation:
  - couch0,couch1,couch2,couch3,couch4,couch5:3:commcarehq__users
  - couch0,couch1,couch2,couch3,couch4,couch5:3
```
The second one will notice `commcarehq__users` was already mentioned and exclude it that piece of the allocation.

**Comment by dannyroberts** [comment]
OK I used this for a series of migrations to move shards off couch[0-3], and it went great. (As you can see I had to make corrections to a couple logical errors in the algorithm.) I'm happy with this and will merge pending your review @snopoke

File: couchdb_cluster_admin/suggest_shard_allocation.py [hunk 1]
```diff
@@ -1,25 +1,152 @@
...
+
+    def _add_shard_to_node(self, node, shard):
+        node.shards.append(shard)
+        node.size += self._sizes_by_shard[shard]
+
+    def _rebalance_nodes(self):
+        larger_nodes, smaller_nodes = self._split_nodes_by_under_allocated()
+        if not smaller_nodes:
+            return
+        print larger_nodes, smaller_nodes
```
Comment (snopoke) [inline couchdb_cluster_admin/suggest_shard_allocation.py]: is this print intentional? consider making it a log statement with more context?

Comment (dannyroberts) [reply couchdb_cluster_admin/suggest_shard_allocation.py]: Oops, thanks. Removed

**Pull request merged**